bpo-34838: Perform a more easily branch-predictable type check

ammaraskar opened the pull request:
More details available on the bpo page.

<!-- issue-number: [bpo-34838](https://www.bugs.python.org/issue34838) -->
https://bugs.python.org/issue34838
<!-- /issue-number -->

MojoVampire commented on Tools/clinic/clinic.py:
```diff
@@ -2727,18 +2725,38 @@ class object_converter(CConverter):
+        if self.subclass_of == '&PyTuple_Type':
+            err_check = '!PyTuple_Check({})'.format(self.name)
+        elif self.subclass_of == '&PyDict_Type':
+            err_check = '!PyDict_Check({})'.format(self.name)
```
Why only tuple and dict? Same fast path would apply to all types with fast subclass flags from object.h ( https://github.com/python/cpython/blob/e42b705188271da108de42b55d9344642170aa2b/Include/object.h#L660 ), including int, list, bytes, str, BaseException and type.

MojoVampire replied:
Admittedly, most things that accept int should probably accept anything convertable to int via `__index__` even if it's not a direct subclass, but otherwise it seems like all the others should have the optimization available.

ammaraskar committed:
bpo-34838: Use subclass_of for math.dist

Argument clinic now generates fast inline code for
positional parsing, so the manually implemented type
check in math.dist can be removed.
